### PR TITLE
A11Y: Improve create account modal for screen readers

### PR DESCRIPTION
**Pull request by martin-brennan**
Improves the create account modal for screen readers by doing the following:

* Making the `modal-alert` section into an `aria-role="alert"` region and making it show and hide using height instead of display:none so screen readers pick it up. Made a change so the field-related error messages are always shown beneath the field.
* Add `aria-invalid` and `aria-describedby` attributes to each field in the modal, so the screen reader will read out the error hint on error. This necessitated an Ember component extension to allow both the `aria-*` attributes to be bound and to render on `{{input}}`.
* Moved the social login buttons to the right in the HTML structure so they are not read out first.
* Added `aria-label` attributes to the login buttons so they can have different content for screen readers.
* In some cases for modals, the title that should be used for the `aria-labelledby` attribute is within the modal content and not the discourse-modal-title title. This introduces a new titleAriaElementId property to the d-modal component that is then used by the create-account modal to read out the

### CODE REVIEW

File: app/assets/javascripts/discourse/app/components/d-modal-body.js [hunk 1]
```diff
@@ -8,7 +8,10 @@ export default Component.extend({
 
   didInsertElement() {
     this._super(...arguments);
-    $("#modal-alert").hide();
+
+    if (!this.preventModalAlertHiding) {
+      $("#modal-alert").hide();
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-modal-body.js:13]: can you try to remove jquery usage here while you are at it please ?

File: app/assets/javascripts/discourse/app/components/d-modal-body.js [hunk 2]
```diff
@@ -70,9 +79,12 @@ export default Component.extend({
   _flash(msg) {
     this._clearFlash();
 
-    $("#modal-alert")
+    let modalAlert = $("#modal-alert")
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/components/d-modal-body.js:82]: same here, would probably be easy to ditch jquery

File: app/assets/javascripts/discourse/app/templates/modal/create-account.hbs [hunk 3]
```diff
@@ -34,8 +31,10 @@
               </div>
 
               <div class="input-group">
-                {{input value=accountUsername disabled=usernameDisabled class=(value-entered accountUsername) id="new-account-username" name="username" maxlength=maxUsernameLength
-                autocomplete="discourse"}}
+                {{input value=accountUsername disabled=usernameDisabled class=(value-entered accountUsername)
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/templates/modal/create-account.hbs:34]: can you format this properly while you are at it please ?

File: app/assets/javascripts/discourse/app/templates/modal/create-account.hbs [hunk 4]
```diff
@@ -78,15 +78,18 @@
 
               <div class="input-group">
                 {{#if passwordRequired}}
-                  {{password-field value=accountPassword class=(value-entered accountPassword) type="password" id="new-account-password" autocomplete="current-password" capsLockOn=capsLockOn}}
+                  {{password-field value=accountPassword class=(value-entered accountPassword)
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/templates/modal/create-account.hbs:81]: format

File: app/assets/javascripts/discourse/tests/acceptance/create-account-user-fields-test.js [hunk 5]
```diff
@@ -39,9 +38,8 @@ acceptance("Create Account - User Fields", function (needs) {
     assert.ok(exists(".user-field"), "it has at least one user field");
 
     await click(".modal-footer .btn-primary");
-    assert.ok(exists("#modal-alert"), "it shows the required field alert");
     assert.equal(
-      queryAll("#modal-alert").text(),
+      queryAll("#account-email-validation").text().trim(),
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/tests/acceptance/create-account-user-fields-test.js:42]: why queryAll and not query ?

File: config/locales/client.en.yml [hunk 6]
```diff
@@ -1897,21 +1897,27 @@ en:
       google_oauth2:
         name: "Google"
         title: "with Google"
+        a11y_title: "Login with Google"
```
Comment (jjaffeux) [inline config/locales/client.en.yml:1900]: I think I would prefer something like: sr_title or screen_reader_title

File: app/assets/stylesheets/common/base/login.scss [hunk 7]
```diff
@@ -57,10 +57,9 @@
     width: 100%;
     max-width: 100%;
     margin-bottom: 0;
-    min-height: 35px;
-  }
-  #modal-alert:empty {
-    display: none;
+    min-height: 50px;
+    position: absolute;
+    z-index: z("modal", "tooltip");
```
Comment (pmusaraj) [inline app/assets/stylesheets/common/base/login.scss:62]: I think this absolute positioning here will cause issues: on mobile, on other locales (for example if the message exceeds the vertical space afforded by the 6em padding) or on other modals that don't have the top padding. 

Can you keep the `:empty` pseudoelement, but instead of hiding with `display: none` or `visibility: hidden`, maybe try something like this: 

```
  #modal-alert {
    display: block;
    padding: 1em 3.5em 1em 1em;
    min-height: 35px;
    ...
    &:empty {
      min-height: 0px;
      padding: 0px;
      overflow: hidden;
    }
  }
```

I don't have NVDA handy, but in a regular browser that seems to behave correctly.

**Comment by jjaffeux** [comment]
@martin-brennan did you consider using this https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques/Using_the_alert_role ?

File: app/assets/javascripts/discourse/app/components/d-modal.js [hunk 8]
```diff
@@ -41,9 +41,17 @@ export default Component.extend({
...
+  @discourseComputed("title", "titleAriaElementId")
+  ariaLabelledby(title, titleAriaElementId) {
+    if (titleAriaElementId) {
+      return titleAriaElementId;
+    }
+    if (title) {
+      return "discourse-modal-title";
+    }
+
+    return null;
```
Comment (eviltrout) [inline app/assets/javascripts/discourse/app/components/d-modal.js:53]: This return is unecessary (unless you specifically need `null` vs `undefined` which would be unusual!)

File: app/assets/javascripts/discourse/app/initializers/ember-input-component-extension.js [hunk 9]
```diff
@@ -0,0 +1,9 @@
+export default {
+  name: "ember-input-component-extensions",
+
+  initialize() {
+    Ember.TextSupport.reopen({
```
Comment (eviltrout) [inline app/assets/javascripts/discourse/app/initializers/ember-input-component-extension.js:5]: This should be imported. `Ember` as a global will be removed.

File: app/assets/javascripts/discourse/tests/acceptance/create-account-user-fields-test.js (hunk 5 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/tests/acceptance/create-account-user-fields-test.js:42]: No idea, was there in the test already, will change

File: app/assets/stylesheets/common/base/login.scss (hunk 7 above)
Comment (martin-brennan) [reply app/assets/stylesheets/common/base/login.scss:62]: Sure thanks for the suggestion, I didn't really know how to do it well so will see if this works

**Comment by martin-brennan** [comment]
@jjaffeux Ah yeah I think you are right with `role="alert"`; it is a live region with an implicit `aria-live` value of `assertive` and an `aria-atomic` value of `true`:

* https://www.w3.org/TR/wai-aria-1.1/#alert
* https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions#roles_with_implicit_live_region_attributes

I will change this, and we should update our internal topic with guidance around using this.

File: config/locales/client.en.yml (hunk 6 above)
Comment (martin-brennan) [reply config/locales/client.en.yml:1900]: Sure, will change this, and we should be consistent with this internally

**Comment by martin-brennan** [comment]
@jjaffeux, @pmusaraj, @eviltrout can you please take a look again? aria-role="alert" worked great and so did Penar's suggestions to use height: 0; so now we have the show/hide of the error as before but now it works with the screen reader.

I grouped the fixes in logical commits, so hopefully should be easy to follow :+1:

File: app/assets/javascripts/discourse/app/controllers/create-account.js [hunk 10]
```diff
@@ -140,16 +140,19 @@ export default Controller.extend(
...
+      "rejectedEmails.[]",
+      "forceValidationReason"
     )
     emailValidation(
       serverAccountEmail,
       serverEmailValidation,
       email,
-      rejectedEmails
+      rejectedEmails,
+      forceValidationReason
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/controllers/create-account.js:151]: having that many computed properties is a smell.

Can you think of any way to group some of them into one property we could use here ?

File: app/assets/javascripts/discourse/app/mixins/password-validation.js [hunk 11]
```diff
@@ -33,18 +33,21 @@ export default Mixin.create({
...
   )
   passwordValidation(
     password,
     passwordRequired,
     rejectedPasswords,
     accountUsername,
     accountEmail,
-    passwordMinLength
+    passwordMinLength,
+    forceValidationReason
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/mixins/password-validation.js:46]: same comment about the number of properties, could be in another PR

File: app/assets/javascripts/discourse/app/templates/modal/create-account.hbs [hunk 12]
```diff
@@ -22,7 +17,16 @@
...
+                {{input type="email"
+                        disabled=emailDisabled
+                        value=accountEmail
+                        id="new-account-email"
+                        name="email"
+                        class=(value-entered accountEmail)
+                        autofocus="autofocus"
+                        focusOut=(action "checkEmailAvailability")
+                        aria-describedby="account-email-validation"
+                        aria-invalid=emailValidation.failed}}
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/templates/modal/create-account.hbs:29]: I don't want to annoy you too much with this but FTR, proper formating is:


```
{{input
  foo=bar
  baz=fooz
}}
``

Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/templates/modal/create-account.hbs:29]: Ah thanks I wasn't sure. You don't have a tool to autoformat these do you?

File: app/assets/javascripts/discourse/app/mixins/password-validation.js (hunk 11 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/mixins/password-validation.js:46]: I agree this is a lot of params, and I would do this in another PR, but what would you suggest doing with these? This and the other computed property are doing complex validations across multiple fields. I don't know how I could group them into one property, and this password validation mixin is used in a couple of different places.

What would your approach be if you had to do it?

Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/mixins/password-validation.js:46]: You could try to group things in state objects:

```
const options = {
  passwordRequired,
  passwordMinLength,
  rejectedPasswords
}

const account = {
  email,
  username
}
```

File: app/assets/javascripts/discourse/app/mixins/password-validation.js [hunk 13]
```diff
@@ -33,18 +33,21 @@ export default Mixin.create({
     "rejectedPasswords.[]",
     "accountUsername",
     "accountEmail",
-    "passwordMinLength"
+    "passwordMinLength",
+    "forceValidationReason"
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/mixins/password-validation.js:37]: Just from reading it, I don't have much idea what this is doing, can we thing of a better name ?

File: app/assets/javascripts/discourse/app/models/login-method.js [hunk 14]
```diff
@@ -13,6 +13,11 @@ const LoginMethod = EmberObject.extend({
     return this.title_override || I18n.t(`login.${this.name}.title`);
   },
 
+  @discourseComputed
+  a11yTitle() {
```
Comment (jjaffeux) [inline app/assets/javascripts/discourse/app/models/login-method.js:17]: maybe we go with `screenReaderTitle` now ?

File: app/assets/javascripts/discourse/app/mixins/password-validation.js (hunk 13 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/mixins/password-validation.js:37]: I had to introduce this because of:

https://github.com/discourse/discourse/blob/0bdda07e336922f957355aa310225623dea8d3a1/app/assets/javascripts/discourse/app/controllers/create-account.js#L163

Some fields should fail silently and not display the reason message under certain conditions. For example, if the password field is empty, we do not want to show the "please enter a password" message straight away, only on submit. So I made this property when submitting the create account form to change the validation behaviour and show the reason always:

https://github.com/discourse/discourse/blob/0bdda07e336922f957355aa310225623dea8d3a1/app/assets/javascripts/discourse/app/controllers/create-account.js#L435

File: app/assets/javascripts/discourse/app/models/login-method.js (hunk 14 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/models/login-method.js:17]: Yes sorry missed that; I think I replaced `a11y_` in this file and not just `a11y`

File: app/assets/javascripts/discourse/app/templates/modal/create-account.hbs (hunk 12 above)
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/templates/modal/create-account.hbs]: I do have one :)

```
find . -name '*.hbs' | xargs yarn prettier --write --parser=glimmer
```
Now I see your question, why dont we use it? Well we can talk of it on chat if you want ;)

File: app/assets/javascripts/discourse/app/mixins/password-validation.js (hunk 13 above)
Comment (jjaffeux) [reply app/assets/javascripts/discourse/app/mixins/password-validation.js:37]: Yes I see.. It's probably related to the fact we use observers in accounts and it creates this kind of issues indeed. I doubt you can do better without heavy refactoring.

**Review by jjaffeux — APPROVED** [review]

File: app/assets/javascripts/discourse/app/mixins/password-validation.js (hunk 11 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/mixins/password-validation.js:46]: Noting for posterity I am going to look into this in another PR, as it will probably require some refactoring.

**Review by pmusaraj — APPROVED** [review]
Looks good to me, thanks!

File: app/assets/javascripts/discourse/app/mixins/password-validation.js (hunk 11 above)
Comment (martin-brennan) [reply app/assets/javascripts/discourse/app/mixins/password-validation.js:46]: After taking a look at this I do not think it is worth the effort of refactoring it right now. Take `emailValidation`; it relies on `accountEmail` along with a few other properties, so then I put those into an `emailData` object the computed property ends up looking like this anyway, so I am not sure why this is better?

```javascript
    // Check the email address
    @discourseComputed(
      "emailData.{accountEmail,rejectedEmails,serverEmailValidation,serverAccountEmail}",
      "forceValidationReason"
    )
    emailValidation(emailData, forceValidationReason) {
```

`accountEmail` is then referenced by username-validation and password-validation, as these mixins are used together, which in turn is then used by both create-account.hbs and invites-show.hbs. invites-show.hbs also uses its own `emailValidation` function which copies a lot but not all of the stuff from create-account.js, further complicating things.

Just seems like a lot of churn for not much benefit to me, so I am leaving this for now. @jjaffeux
